src/op_handlers/ret.rs

```diff
@@ -0,0 +1,79 @@
+use u256::U256;
+use zkevm_opcode_defs::RetOpcode;
+
+use crate::{
+    eravm_error::EraVmError,
+    state::VMState,
+    store::Storage,
+    value::{FatPointer, TaggedValue},
+    Opcode,
+};
+
+use super::far_call::get_forward_memory_pointer;
+
+fn is_failure(return_type: RetOpcode) -> bool {
+    return_type != RetOpcode::Ok
+}
+
+fn get_result(
+    vm: &mut VMState,
+    opcode: &Opcode,
+    return_type: RetOpcode,
+) -> Result<TaggedValue, EraVmError> {
+    if return_type == RetOpcode::Panic {
+        return Ok(TaggedValue::new_pointer(U256::zero()));
+    }
+    let register = vm.get_register(opcode.src0_index);
+    let result = get_forward_memory_pointer(register.value, vm, register.is_pointer)?;
+    Ok(TaggedValue::new_pointer(FatPointer::encode(&result)))
+}
+
+pub fn ret(
+    vm: &mut VMState,
+    opcode: &Opcode,
+    storage: &mut dyn Storage,
+    return_type: RetOpcode,
+    use_label: bool,
+) -> Result<bool, EraVmError> {
+    let is_failure = is_failure(return_type);
+
+    vm.flag_eq = false;
+    vm.flag_lt_of = return_type == RetOpcode::Panic;
+    vm.flag_gt = false;
+
+    if is_failure {
+        storage.rollback(&vm.current_frame()?.storage_before);
+    }
+
+    if vm.in_near_call()? {
+        let previous_frame = vm.pop_frame()?;
+        if use_label && opcode.alters_vm_flags {
+            let to_label = opcode.imm0;
+            vm.current_frame_mut()?.pc = to_label as u64;
```

Comment:
The opcode should only be used if this is an actual Ok, Revert or Panic, if this was due to out of gas or a vm error, then the opcode could be anything and we don't want to use its data